lib/mongo/srv/resolver.rb

```diff
@@ -0,0 +1,65 @@
+# Copyright (C) 2014-2019 MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'resolv'
+
+module Mongo
+  module SRV
+
+    # The SRV::Resolver class encapsulates the necessary behavior for querying SRV records.
+    #
+    # @api private
+    class Resolver
+      include Loggable
+
+      RECORD_PREFIX = '_mongodb._tcp.'.freeze
+
+      def initialize(raise_on_invalid = false)
+        @raise_on_invalid = raise_on_invalid
+        @resolver = Resolv::DNS.new
+      end
+
+      def get_records(hostname)
```

Comment:
Is a trailing dot in hostname allowed, not allowed, required?

Reply:
I don't think the spec says anything about this, and I don't understand what that question has to do with the section of the code you commented on

Reply:
The question has to do with how DNS works. DNS records always have trailing dots, e.g. `www.mongodb.com.`. Virtually in all other usage trailing dots are omitted, i.e. we talk about `www.mongodb.com`. Since the code in question deals with DNS records and performs hostname matching, which splits the hostnames on dots, it is important that trailing dots be correctly handled. It is also important for calling code to know whether `hostname` should or should not have the trailing dot.

Reply:
All of the code related to parsing the hostname and doing the DNS queries is copied from the existing implementation in `mongo/uri/srv_protocol.rb`; all this PR does is factor it out into separate classes. I think it might be useful for you to read the existing implementation so you understand what's actually being changed here.